core/src/heaviest_subtree_fork_choice.rs

```diff
@@ -411,6 +418,59 @@ impl HeaviestSubtreeForkChoice {
             .map(|(slot_hash, fork_info)| (slot_hash, fork_info.stake_voted_subtree))
     }
 
+    /// Dump the node `slot_hash_key` and propagate the stake subtraction up to the root of the
+    /// tree. Children of `slot_hash_key` are implicitely dumped as well, as if we were able to
+    /// chain them to a defunct parent that implies that they are defunct as well (consistent with
+    /// bank forks).
+    ///
+    /// Returns all removed slots
+    pub fn dump_node(&mut self, slot_hash_key: &SlotHashKey) -> Vec<Slot> {
+        let parent = {
+            let mut node_to_dump = self
+                .fork_infos
+                .get_mut(slot_hash_key)
+                .expect("Slot hash key must exist in tree");
```

Comment:
Is there a reason we can't delete these nodes if they're dumped?

Reply:
we delete these nodes later in the function (line 451). Before deleting these nodes we 0 their stake so we can reuse the aggregation functions to modify stake / best child info to ancestors.

Reply:
ah, missed this. Thanks for the explainer!